pkg/common/git.go

```diff
@@ -0,0 +1,192 @@
+package common
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/nuclio/nuclio/pkg/cmdrunner"
+
+	"github.com/nuclio/errors"
+	"github.com/nuclio/logger"
+	"gopkg.in/src-d/go-git.v4"
+	"gopkg.in/src-d/go-git.v4/plumbing"
+	githttp "gopkg.in/src-d/go-git.v4/plumbing/transport/http"
+)
+
+type GitAttributes struct {
+	Branch    string `json:"branch,omitempty"`
+	Tag       string `json:"tag,omitempty"`
+	Reference string `json:"reference,omitempty"`
+	Username  string `json:"username,omitempty"`
+	Password  string `json:"password,omitempty"`
+}
+
+type GitClient interface {
+	Clone(outputDir, repositoryURL string, gitAttributes *GitAttributes) error
+}
+
+type AbstractGitClient struct {
+	GitClient
+
+	logger    logger.Logger
+	cmdRunner cmdrunner.CmdRunner
+}
+
+func NewGitClient(parentLogger logger.Logger) (GitClient, error) {
+	var err error
+
+	abstractGitClient := AbstractGitClient{logger: parentLogger.GetChild("git-client")}
+
+	// create cmd runner
+	abstractGitClient.cmdRunner, err = cmdrunner.NewShellRunner(parentLogger)
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to create cmd runner")
+	}
+
+	return abstractGitClient, nil
+}
+
+func (agc AbstractGitClient) Clone(outputDir, repositoryURL string, gitAttributes *GitAttributes) error {
+	var referenceName string
+	var gitAuth *githttp.BasicAuth
+	var err error
+
+	// resolve full git reference name
+	referenceName, err = ResolveGitReference(repositoryURL, gitAttributes)
+	if err != nil {
+		return errors.Wrap(err, "Failed to resolve git reference")
+	}
+
+	// resolve git credentials when given
+	gitAuth = agc.parseFunctionGitCredentials(gitAttributes)
+
+	// if it's Azure Devops repo - clone differently (the normal go-git client doesn't support it yet)
+	if isAzureDevopsRepositoryURL(repositoryURL) {
+		return agc.cloneFromAzureDevops(outputDir, repositoryURL, referenceName, gitAuth, agc.cmdRunner)
+	}
+
+	if _, err = git.PlainClone(outputDir, false, &git.CloneOptions{
+		URL:           repositoryURL,
+		ReferenceName: plumbing.ReferenceName(referenceName),
+		Depth:         1,
+		Auth:          gitAuth,
+	}); err != nil {
+		return errors.Wrap(err, "Failed to clone git repository")
+	}
+
+	agc.printCurrentCommitSHA(outputDir, repositoryURL, referenceName)
+
+	return nil
+}
+
+func (agc AbstractGitClient) printCurrentCommitSHA(gitDir, referenceName, repositoryURL string) {
+	res, err := agc.cmdRunner.Run(nil, fmt.Sprintf("cd %s;git rev-parse HEAD", Quote(gitDir)))
```

Comment:
Two things here
1. you can use `RunOptions` and set up the `workingDir`
2. You have a git client - use that to resolve current git sha

The idea, is to later on remove the need of command runner, once AzureDevop would be fully supported